dwave/gate/simulator/simulator.pyx

```diff
@@ -145,30 +153,106 @@ def simulate(
     state[0] = 1
 
     for op in circuit.circuit:
-        targets = [circuit.qubits.index(qb) for qb in op.qubits]
-        apply_instruction(num_qubits, state, op, targets, little_endian)
+        if isinstance(op, ops.Measurement):
+            _measure(op, state, circuit)
+        else:
+            targets = [circuit.qubits.index(qb) for qb in op.qubits]
+            apply_instruction(num_qubits, state, op, targets, little_endian)
 
     return state
 
 
+def _measure(op, state, circuit):
+    op._measured_state = state.copy()
+
+    for idx, qb in enumerate(op.qubits):
+        m = _sample(circuit.qubits.index(qb), state)
+
+        try:
+            op.bits[idx].set(m)
+        except (IndexError, TypeError):
+            warnings.warn("Measurements not stored in the classical register.")
+
+        op._measured_qubit_indices.append(circuit.qubits.index(qb))
+
+
+def _sample(
+    qubit: int,
+    state: np.typing.NDArray,
+    collapse_state: bool = True,
+    reset_measured_qubits: bool = False
+) -> int:
+    """Sample a measurement on a single qubit and collapse the state.
+
+    Args:
+        qubit: Which qubit (index, from left to right) to sample.
+        collapse_state: Whether to collapse the state after a measurement.
+        reset_measured_qubits: Whether to reset the measured qubits to 0 after being measured.
+
+    Returns:
+        int: 0 or 1 sample of the measured qubit.
+    """
+    num_qubits = round(np.log2(len(state)))
+
+    if qubit >= num_qubits:
+        raise ValueError(f"Cannot sample qubit {qubit}. State has only {num_qubits} qubits.")
+
+    weight_0 = weight_1 = 0
+    zero_weight = False
+    state_0_ids = []
+    state_1_ids = []
+
+    for i in range(2**num_qubits):
+        if i % 2**(num_qubits - qubit - 1) == 0:
+            zero_weight = not zero_weight
+        if zero_weight:
```

Comment:
Something like `if (i >> (num_qubits - qubit - 1) & 1:` may be much faster here (and similar the `sample` function in the other file). Modulo on every loop is expensive